internal/graph/check.go

```diff
@@ -0,0 +1,612 @@
+}
+
+type setOperatorType int
+
+const (
+	unionSetOperator setOperatorType = iota
+	intersectionSetOperator
+	exclusionSetOperator
+)
+
+type checkOutcome struct {
+	resp *openfgapb.CheckResponse
+	err  error
+}
+
+// LocalChecker implements Check in a highly concurrent and localized manner. The
+// Check resolution is limited per branch of evaluation by the concurrencyLimit.
+type LocalChecker struct {
+	ds               storage.RelationshipTupleReader
+	concurrencyLimit uint32
+}
+
+// NewLocalChecker constructs a LocalChecker that can be used to evaluate a Check
+// request locally and with a high degree of concurrency.
+func NewLocalChecker(
+	ds storage.RelationshipTupleReader,
+	concurrencyLimit uint32,
+) *LocalChecker {
+	checker := &LocalChecker{ds: ds, concurrencyLimit: concurrencyLimit}
+	return checker
+}
+
+// CheckHandlerFunc defines a function that evaluates a CheckResponse or returns an error
+// otherwise.
+type CheckHandlerFunc func(ctx context.Context) (*openfgapb.CheckResponse, error)
+
+// CheckFuncReducer defines a function that combines or reduces one or more CheckHandlerFunc into
+// a single CheckResponse with a maximum limit on the number of concurrent evaluations that can be
+// in flight at any given time.
+type CheckFuncReducer func(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error)
+
+// resolver concurrently resolves one or more CheckHandlerFunc and yields the results on the provided resultChan.
+// Callers of the 'resolver' function should be sure to invoke the callback returned from this function to ensure
+// every concurrent check is evaluated. The concurrencyLimit can be set to provide a maximum number of concurrent
+// evaluations in flight at any point.
+func resolver(ctx context.Context, concurrencyLimit uint32, resultChan chan<- checkOutcome, handlers ...CheckHandlerFunc) func() {
+	limiter := make(chan struct{}, concurrencyLimit)
+
+	var wg sync.WaitGroup
+
+	checker := func(fn CheckHandlerFunc) {
+		resp, err := fn(ctx)
+		resultChan <- checkOutcome{resp, err}
+		<-limiter
+		wg.Done()
+	}
+
+	wg.Add(1)
+	go func() {
+	outer:
+		for _, handler := range handlers {
+			fn := handler // capture loop var
+
+			select {
+			case limiter <- struct{}{}:
+				wg.Add(1)
+				go checker(fn)
+			case <-ctx.Done():
+				break outer
+			}
+		}
+
+		wg.Done()
+	}()
+
+	return func() {
+		wg.Wait()
+		close(limiter)
+	}
+}
+
+// union implements a CheckFuncReducer that requires any of the provided CheckHandlerFunc to resolve
+// to an allowed outcome. The first allowed outcome causes premature termination of the reducer.
+func union(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error) {
+
+	ctx, cancel := context.WithCancel(ctx)
+	resultChan := make(chan checkOutcome, len(handlers))
+
+	drain := resolver(ctx, concurrencyLimit, resultChan, handlers...)
+
+	defer func() {
+		cancel()
+		drain()
```

Comment:
As you found offline, this will wait until all functions finish, which is not what we want.

Reply:
Ok, I've fixed the issue now. I've enabled to the two tests in `pkg/server/server_test.go` (`TestCheckDoesNotThrowBecauseDirectTupleWasFound` and `TestShortestPathToSolutionWins`) to demonstrate that we a) don't return an error if another resolution path returned before the error and b) short-circuit if a resolution path was found (in the union case) or if a resolution path was rejected (in the exclusion and intersection cases) prior to other check function handlers resolve.

Reply:
The changes also have little/to no impact on the overall goroutine footprint as well. The following charts show a 2m load tests (250rps of random Checks involving direct, computed_userset, and ttu resolutions). The first chart is this refactored Check and the second chart shows the results before the refactor.

<img width="1906" alt="Screen Shot 2023-02-06 at 7 31 15 PM" src="https://user-images.githubusercontent.com/2899204/217135140-bfeace0f-da87-4fa0-9b15-7332af9b1d80.png">

<img width="1894" alt="Screen Shot 2023-02-06 at 7 37 53 PM" src="https://user-images.githubusercontent.com/2899204/217135216-527fc49a-6b7f-4ff4-9cd6-111b1cfc2b38.png">

Reply:
Can run a longer test of a few hours?

Reply:
Sure!

Reply:
I ran the k6-load-test script last night for 4 hours to check for memory leaks etc.. Everything looked really good. No signs of memory leaks or excessive processing.

The test run started just before 4:00.

<img width="1696" alt="Screen Shot 2023-02-07 at 7 38 34 AM" src="https://user-images.githubusercontent.com/2899204/217281637-0fc9b7ae-a776-4f01-916d-90c8c53cdbec.png">
<img width="1694" alt="Screen Shot 2023-02-07 at 7 38 53 AM" src="https://user-images.githubusercontent.com/2899204/217281670-1bf5662c-bdf0-4eee-9d84-4f8433e37245.png">
<img width="1700" alt="Screen Shot 2023-02-07 at 7 39 03 AM" src="https://user-images.githubusercontent.com/2899204/217281704-d9fecb26-7ca4-41d9-8320-fd984c9d4e4c.png">